src/codewhisperer/commands/basicCommands.ts

```diff
@@ -120,37 +121,41 @@ export const showSsoSignIn = Commands.declare('aws.codeWhisperer.sso', () => asy
-        async (
-            startUrl?: string,
-            region?: string,
-            customizationArn?: string,
-            customizationName?: string,
-            customizationDescription?: string
-        ) => {
-            // This command supports two sets of arguments:
-            //  * startUrl and region. If both arguments are provided they will be used, otherwise
-            //    the command prompts for them interactively.
-            //  * customization{Arn, Name, Description}. If at least customizationArn is provided,
-            //    the command selects this customization.
-            if (startUrl && region) {
-                await connectToEnterpriseSso(startUrl, region)
-            } else {
-                await getStartUrl()
-            }
-            if (customizationArn) {
-                const match = getPersistedCustomizations().find(c => c.arn == customizationArn)
-                const customization = {
-                    arn: customizationArn,
-                    name: customizationName ?? match?.name ?? 'unknown',
-                    description: customizationDescription ?? match?.description ?? 'unknown',
-                }
-                await selectCustomization(customization)
-            }
+    () => async (startUrl?: string, region?: string, customizationArn?: string, customizationNamePrefix?: string) => {
+        // This command supports the following arguments:
+        //  * startUrl and region. If both arguments are provided they will be used, otherwise
+        //    the command prompts for them interactively.
+        //  * customizationArn: select customization by ARN. If provided, `customizationNamePrefix` is ignored.
+        //  * customizationNamePrefix: select customization by prefix, if `customizationArn` is `undefined`.
+        if (startUrl && region) {
+            await connectToEnterpriseSso(startUrl, region)
+        } else {
+            await getStartUrl()
         }
+
+        // No customization match information given, exit early.
+        if (!customizationArn && !customizationNamePrefix) {
+            return
+        }
+
+        // If given an ARN, assume a specific customization is desired and find an entry that matches it. Ignores the prefix logic.
+        // Otherwise if only a prefix is given, find an entry that matches it.
+        // Backwards compatible with previous implementation.
+        const match = customizationArn
+            ? getPersistedCustomizations().find(c => c.arn === customizationArn)
+            : getPersistedCustomizations().find(c => c.name?.startsWith(customizationNamePrefix as string))
+
+        // If no match is found, nothing to do :)
+        if (!match) {
+            getLogger().error(`No customization match found: arn=${customizationArn} prefix=${customizationNamePrefix}`)
+            return
+        }
+        // Since we selected based on a match, we'll reuse the persisted values.
+        await selectCustomization(match)
```

Comment:
I think this would actually cause some regression comparing to the current behavior, for the new users.
For the new users, there're no persisted customizations on their machine, so match would return nothing. In this case it will just return, in the previous logic you continued to use the arn the caller specified which actually will work if it matches the arn the service side has.

Reply:
To rephrase, the problem now becomes: "How does it auto-select the customizations for new users if they have't used customizations before?"
Could you manually test this case?

Reply:
During testing, we saw that persisted customizations are present after establishing a connection, which is the first action we take before we start attempting to find a match.

In manual testing, whether we gave it just the ARN or just the prefix, we saw it successfully set the correct customization. We also executed `setPersistedCustomizations([])` to clear out any stored entries prior to executing the command. Lastly, we also used a different prefix (`NGDE`) to make sure we used a customization that hasn't been used/set yet.

![image](https://github.com/aws/aws-toolkit-vscode/assets/13372878/2b306e43-ad9b-47c4-90c0-392c5d9e76b3)

All in all, it seems to be doing what we expected, but I'm happy to follow up on any additional testing to make sure it's as robust as possible.

Reply:
Did you call `await setPersistedCustomizations([])` right after `await connectToEnterpriseSso(startUrl, region)`?

The code doesn't look like it will handle this case, so it's suprising to me that this is still passing.

Reply:
We did call `await setPersistedCustomizations([])`, but _before_ the connection logic.

I'm not quite sure why we would want to test the case where we clear the persisted customizations after establishing the connection when the logic is designed to execute _only if_ a connection has been established.

In the worst case, say if a connection is not established, then it will not set a customization as expected.

Reply:
The test case is for when a new user has installed Viceroy and CodeWhisperer for the first time, thus their machine does not have any info about any persisted customization, calling `await setPersistedCustomizations([])` before or after connection logic doesn't really matter, as long as it's cleared before we call `getPersistedCustomization()` it would simulate the usecase.

The worst case is when a connection _is_ established but we don't set the customization as expected. And that's beacuse we try to match persisted ones but users don't have such on their machine, comparing to the existing code that prioritizes the passed parameter. This is making the behavior to change.

existing:
```
if (customizationArn) {
                const match = getPersistedCustomizations().find(c => c.arn == customizationArn)
                const customization = {
                    arn: customizationArn, <-- here means even if match doesn't return anything we still favor the given arn
                    name: customizationName ?? match?.name ?? 'unknown',
                    description: customizationDescription ?? match?.description ?? 'unknown',
                }
                await selectCustomization(customization)
            }
```

```
const match = customizationArn
            ? getPersistedCustomizations().find(c => c.arn === customizationArn)
            : getPersistedCustomizations().find(c => c.name?.startsWith(customizationNamePrefix as string))

        // If no match is found, nothing to do :)
        if (!match) { <-- here it means if we don't find a match we will return early without setting any customization.
            getLogger().error(`No customization match found: arn=${customizationArn} prefix=${customizationNamePrefix}`)
            return
```

Reply:
As we discussed offline, it seems the use of persisted customizations might encounter a race condition. The expected behavior is that it shouldn't be set by establishing the connection alone and might only be set as a side-effect.

We'll add a check first to make sure the list isn't empty and call on the notify logic to establish the persisted list.